File: processing/src/ipc/control_socket.py
"""
Control Socket IPC for configuration, telemetry, and live transcription events.
"""

import socket
import struct
import json
import threading
from typing import Callable, Optional, List


class ControlSocketClient:
    def __init__(self, host: str = "127.0.0.1", port: int = 9200):
        self.host = host
        self.port = port
        self.sock: Optional[socket.socket] = None
        self.running = False
        self.on_message_callback: Optional[Callable[[dict], None]] = None
        self._recv_thread: Optional[threading.Thread] = None

    def connect(self) -> bool:
        try:
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.sock.connect((self.host, self.port))
            self.running = True
            self._recv_thread = threading.Thread(target=self._listen_loop, daemon=True)
            self._recv_thread.start()
            return True
        except Exception as e:
            print(f"[ControlSocketClient] Connect to {self.host}:{self.port} failed: {e}")
            return False

    def send(self, message: dict) -> bool:
        if not self.sock or not self.running:
            return False
        try:
            payload = json.dumps(message).encode("utf-8")
            length_prefix = struct.pack(">I", len(payload))
            self.sock.sendall(length_prefix + payload)
            return True
        except Exception as e:
            print(f"[ControlSocketClient] Send error: {e}")
            return False

    def _listen_loop(self):
        while self.running and self.sock:
            try:
                raw_len = self.sock.recv(4)
                if not raw_len or len(raw_len) < 4:
                    break
                (length,) = struct.unpack(">I", raw_len)
                data = bytearray()
                while len(data) < length:
                    packet = self.sock.recv(length - len(data))
                    if not packet:
                        break
                    data.extend(packet)

                if len(data) == length:
                    msg = json.loads(data.decode("utf-8"))
                    if self.on_message_callback:
                        self.on_message_callback(msg)
            except Exception as e:
                break
        self.running = False

    def close(self):
        self.running = False
        if self.sock:
            try:
                self.sock.close()
            except Exception:
                pass
            self.sock = None


class ControlSocketServer:
    def __init__(self, host: str = "127.0.0.1", port: int = 9201):
        self.host = host
        self.port = port
        self.server_sock: Optional[socket.socket] = None
        self.clients: List[socket.socket] = []
        self.lock = threading.Lock()
        self.running = False
        self.on_message_callback: Optional[Callable[[dict], None]] = None

    def start(self):
        self.server_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.server_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.server_sock.bind((self.host, self.port))
        self.server_sock.listen(5)
        self.running = True

        thread = threading.Thread(target=self._accept_loop, daemon=True)
        thread.start()
        print(f"[ControlSocketServer] Python Control Server listening on {self.host}:{self.port}")

    def broadcast(self, message: dict):
        payload = json.dumps(message).encode("utf-8")
        msg_bytes = struct.pack(">I", len(payload)) + payload

        with self.lock:
            active = []
            for c in self.clients:
                try:
                    c.sendall(msg_bytes)
                    active.append(c)
                except Exception:
                    pass
            self.clients = active

    def _accept_loop(self):
        while self.running and self.server_sock:
            try:
                client_sock, _ = self.server_sock.accept()
                with self.lock:
                    self.clients.append(client_sock)
                threading.Thread(target=self._client_handler, args=(client_sock,), daemon=True).start()
            except Exception:
                break

    def _client_handler(self, client_sock: socket.socket):
        while self.running:
            try:
                raw_len = client_sock.recv(4)
                if not raw_len or len(raw_len) < 4:
                    break
                (length,) = struct.unpack(">I", raw_len)
                data = bytearray()
                while len(data) < length:
                    packet = client_sock.recv(length - len(data))
                    if not packet:
                        break
                    data.extend(packet)

                if len(data) == length:
                    msg = json.loads(data.decode("utf-8"))
                    if self.on_message_callback:
                        self.on_message_callback(msg)
            except Exception:
                break
        with self.lock:
            if client_sock in self.clients:
                self.clients.remove(client_sock)
        client_sock.close()

    def stop(self):
        self.running = False
        if self.server_sock:
            self.server_sock.close()

File: transmission/src/ipc/control_socket.cpp
#include "control_socket.h"
#include <iostream>
#include <cstring>
#include <vector>

#if defined(_WIN32)
#ifndef WIN32_LEAN_AND_MEAN
#define WIN32_LEAN_AND_MEAN
#endif
#ifndef NOMINMAX
#define NOMINMAX
#endif
#include <winsock2.h>
#include <ws2tcpip.h>
typedef SOCKET socket_t;
#define IS_INVALID_SOCKET(s) ((s) == INVALID_SOCKET)
#define CLOSE_SOCKET(s) closesocket(s)
#else
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
typedef int socket_t;
#define INVALID_SOCKET (-1)
#define IS_INVALID_SOCKET(s) ((s) < 0)
#define CLOSE_SOCKET(s) close(s)
#endif

namespace voip {

ControlSocketServer::ControlSocketServer(uint16_t port)
    : m_port(port)
{
#if defined(_WIN32)
    WSADATA wsaData;
    WSAStartup(MAKEWORD(2, 2), &wsaData);
#endif
}

ControlSocketServer::~ControlSocketServer() {
    stop();
#if defined(_WIN32)
    WSACleanup();
#endif
}

void ControlSocketServer::onMessage(MessageCallback callback) {
    m_callback = std::move(callback);
}

bool ControlSocketServer::start() {
    if (m_running) return true;
    m_running = true;
    m_listenThread = std::thread(&ControlSocketServer::listenLoop, this);
    return true;
}

void ControlSocketServer::stop() {
    if (!m_running) return;
    m_running = false;

    {
        std::lock_guard<std::mutex> lock(m_clientsMutex);
        for (auto sock : m_clientSockets) {
            CLOSE_SOCKET(static_cast<socket_t>(sock));
        }
        m_clientSockets.clear();
    }

    if (m_listenThread.joinable()) {
        m_listenThread.detach(); // or join if unblocked
    }
}

void ControlSocketServer::broadcast(const std::string& jsonMessage) {
    uint32_t len = static_cast<uint32_t>(jsonMessage.size());
    uint32_t netLen = htonl(len);

    std::vector<uint8_t> buffer(4 + len);
    std::memcpy(buffer.data(), &netLen, 4);
    std::memcpy(buffer.data() + 4, jsonMessage.data(), len);

    std::lock_guard<std::mutex> lock(m_clientsMutex);
    std::vector<intptr_t> active;
    for (auto sock : m_clientSockets) {
        int sent = send(static_cast<socket_t>(sock), reinterpret_cast<const char*>(buffer.data()), static_cast<int>(buffer.size()), 0);
        if (sent > 0) {
            active.push_back(sock);
        } else {
            CLOSE_SOCKET(static_cast<socket_t>(sock));
        }
    }
    m_clientSockets = std::move(active);
}

void ControlSocketServer::listenLoop() {
    socket_t serverSock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (IS_INVALID_SOCKET(serverSock)) {
        std::cerr << "[ControlSocketServer] Failed to create server socket." << std::endl;
        m_running = false;
        return;
    }

    int opt = 1;
#if defined(_WIN32)
    setsockopt(serverSock, SOL_SOCKET, SO_REUSEADDR, (const char*)&opt, sizeof(opt));
#else
    setsockopt(serverSock, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));
#endif

    sockaddr_in addr{};
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    addr.sin_port = htons(m_port);

    if (bind(serverSock, (sockaddr*)&addr, sizeof(addr)) != 0) {
        std::cerr << "[ControlSocketServer] Bind failed on port " << m_port << std::endl;
        CLOSE_SOCKET(serverSock);
        m_running = false;
        return;
    }

    if (listen(serverSock, 5) != 0) {
        std::cerr << "[ControlSocketServer] Listen failed." << std::endl;
        CLOSE_SOCKET(serverSock);
        m_running = false;
        return;
    }

    std::cout << "[ControlSocketServer] Listening for IPC control on 127.0.0.1:" << m_port << std::endl;

    while (m_running) {
        sockaddr_in clientAddr{};
        socklen_t clientLen = sizeof(clientAddr);
        socket_t clientSock = accept(serverSock, (sockaddr*)&clientAddr, &clientLen);
        if (IS_INVALID_SOCKET(clientSock)) {
            if (!m_running) break;
            continue;
        }

        {
            std::lock_guard<std::mutex> lock(m_clientsMutex);
            m_clientSockets.push_back(static_cast<intptr_t>(clientSock));
        }

        std::thread([this, clientSock]() {
            this->clientWorker(static_cast<intptr_t>(clientSock));
        }).detach();
    }

    CLOSE_SOCKET(serverSock);
}

void ControlSocketServer::clientWorker(intptr_t clientSockVal) {
    socket_t sock = static_cast<socket_t>(clientSockVal);

    while (m_running) {
        uint32_t netLen = 0;
        int received = recv(sock, reinterpret_cast<char*>(&netLen), 4, MSG_WAITALL);
        if (received <= 0) {
            break;
        }

        uint32_t len = ntohl(netLen);
        if (len > 10 * 1024 * 1024) { // 10MB limit safety
            break;
        }

        std::string payload(len, '\0');
        received = recv(sock, &payload[0], len, MSG_WAITALL);
        if (received <= 0) {
            break;
        }

        if (m_callback) {
            m_callback(payload);
        }
    }

    CLOSE_SOCKET(sock);
}

} // namespace voip
